midterm/src/Directory.hs

```diff
@@ -1,17 +1,48 @@
-module Directory where
+module Directory (DirectoryTree(..), buildTree, defaultDisplayTree, alphabetisedDisplayTree) where
 
-import Tree
+import System.Directory (doesDirectoryExist, listDirectory)
+import System.FilePath ((</>))
+import Data.List (sort)
 
 -- Define an appropriate type called `DirectoryTree` to represent the directory hierarchy
+data DirectoryTree = File FilePath | Directory FilePath [DirectoryTree]
+    deriving (Show, Eq)
 
 -- Create a tree that represents the hierarchy of files and subdirectories of the given directory.
 buildTree :: FilePath -> IO DirectoryTree
-buildTree = undefined 
+buildTree filePath = do
+    check <- doesDirectoryExist filePath
+    if check
+        then do
+            files <- listDirectory filePath
+            let sortedFiles = sort files -- probably sorting by default is more efficient than separating by default(???)
+            sub <- mapM (buildTree . (filePath </>)) sortedFiles
+            return $ Directory filePath sub
+        else
+            return $ File filePath
+
+-- Separate tree
+partitionDirectoriesAndFiles :: [DirectoryTree] -> ([DirectoryTree], [DirectoryTree])
+partitionDirectoriesAndFiles = foldr separate ([], [])
+  where
+    separate item (dirs, files) = case item of
+        Directory _ _ -> (item : dirs, files)
+        File _        -> (dirs, item : files)
+
+-- Function to format the tree
+formatTree :: String -> DirectoryTree -> String
+formatTree indent (File filePath) = indent ++ "- " ++ filePath ++ "\n"
+formatTree indent (Directory filePath files) =
+    indent ++ "+ " ++ filePath ++ "\n" ++ concatMap (formatTree (indent ++ "  ")) files
 
 -- Implement a function that displays the subdirectories of the current directory before the files in it. 
 defaultDisplayTree :: DirectoryTree -> String
-defaultDisplayTree = undefined 
+defaultDisplayTree (File filePath) = "- " ++ filePath ++ "\n"
```

Comment:
This produces an infinite loop. However, when I fixed the infinite loop, it still displayed the tree alphabetically.